### ISSUE TITLE
Returning owned array when function expects a slice with a trait results in misleading error message.

### ISSUE THREAD

**Issue by dogunbound** [rust-lang/rust#109053]
### Code

```Rust
trait MyTrait {}
struct ImT1 {}
impl MyTrait for ImT1 {}
struct ImT2 {}
impl MyTrait for ImT2 {}
pub struct MyStruct {
  imt1: ImT1,
  imt2: ImT2,
}

impl MyStruct {
  pub fn mytrait_array_mut(&mut self) -> &[&dyn MyTrait] {
    [
      self.imt1,
      self.imt2,
    ]
  }  
}
```

### Current output

```Shell
error[E0308]: mismatched types
  --> src/main.rs:16:7
   |
16 |       self.imt2,
   |       ^^^^^^^^^ expected `ImT1`, found `ImT2`

For more information about this error, try `rustc --explain E0308`.
```

### Desired output

```Shell
What should the suggested output be? Leaving it blank and hoping for good recommendations.
```

### Rationale and extra context

This error message is WORSE than a segfault because it is misleading. It gives the wrong message. At least segfaults tell you that there was an illegal memory operation.

**Comment by eddyb** [comment]
> The error message should reference the fact that the return value does not match the function signature.

I'll note that before that mismatch can even be detected, you have the issue of a `[a, b]` expression where `a` and `b` have different types. The only reason that can ever work is *coercion* (which here can't trigger because you don't have references, and the array/slice mismatch, but that's besides the point).

We skip erroring about the return mismatch because the type already has errors in it, but maybe we shouldn't, since `&[_]` and `[_; _]` would just never "fit" together regardless of their substructure (automating this may be hard without 1. replacing all nested errors with new inference variables 2. a way to ask the coercion machinery if whether a type can coerce to another is ambiguous vs impossible).

---

We already (IIRC) track all the expressions involved in a N-way coercion attempt (such as an array expression here, or an `if`-`else`/`match` expression), so we should be able to point at every single expression involved (whose types don't line up) and label them with their types (including maybe some intermediary coercion targets, if any coercions were found), so that the user has a better grasp on where the mismatch is coming from.

For that part of the problem, this is enough to repro:
```rust
fn main() {
    [
        "foo",
        true,
    ];
}
```
```
error[E0308]: mismatched types
 --> src/main.rs:4:9
  |
4 |         true,
  |         ^^^^ expected `&str`, found `bool`
```

There is no indication of where `&str` came from, and it looks kinda context-free, a better message could be:
```
error[E0308]: mismatched types between array elements
 --> src/main.rs:4:9
  |
3 |         "foo",
  |         ^^^^^ expected `&str`
4 |         true,
  |         ^^^^ found `bool`
```
(I kept `expected`/`found` even if I don't like them, because I don't want to bikeshed an improvement here)

---

> This error message is WORSE than a segfault because it is misleading. It gives the wrong message. At least segfaults tell you that there was an illegal memory operation.

You wouldn't even get a segfault reliably with an UAF caused by, say, returning stack memory by reference (you're far more likely to get silent data corruption), so the hyperbole is wrong (on top of being unnecessary).
